File: server/app/schemas/kb_permission_rule.py
"""
KbPermissionRule Pydantic schemas
"""
from pydantic import BaseModel, ConfigDict, Field

from app.enums import UserConditionOperator, ScopeOperator
from app.schemas.base import TimestampSchema


class UserCondition(BaseModel):
    field: str = Field(..., min_length=1, max_length=128, description="metadata key name")
    operator: UserConditionOperator
    value: str | list[str] | None = Field(
        None, description="Comparison value; null for is_empty/is_not_empty"
    )


class KbPermissionRuleCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=128)
    enabled: bool = True
    user_conditions: list[UserCondition] = Field(..., min_length=1)
    scope_operator: ScopeOperator
    scope_labels: list[str] | None = Field(
        None, description="Required for equals/not_equals, null for contains_any/not_contains_any"
    )


class KbPermissionRuleUpdate(BaseModel):
    name: str | None = Field(None, min_length=1, max_length=128)
    enabled: bool | None = None
    user_conditions: list[UserCondition] | None = Field(None, min_length=1)
    scope_operator: ScopeOperator | None = None
    scope_labels: list[str] | None = None


class KbPermissionRuleResponse(TimestampSchema):
    model_config = ConfigDict(from_attributes=True)

    id: int
    tenant_id: str
    knowledge_base_id: int
    name: str
    enabled: bool
    user_conditions: list[dict]
    scope_operator: str
    scope_labels: list[str] | None = None

File: web/models/kb-permission-rule.ts
export type UserConditionOperator =
  | 'equals'
  | 'not_equals'
  | 'contains'
  | 'not_contains'
  | 'starts_with'
  | 'ends_with'
  | 'in'
  | 'not_in'
  | 'is_empty'
  | 'is_not_empty'

export type ScopeOperator =
  | 'equals'
  | 'not_equals'
  | 'contains_any'
  | 'not_contains_any'

export type UserCondition = {
  field: string
  operator: UserConditionOperator
  value: string | string[] | null
}

export type KbPermissionRule = {
  id: number
  tenant_id: string
  knowledge_base_id: number
  name: string
  enabled: boolean
  user_conditions: UserCondition[]
  scope_operator: ScopeOperator
  scope_labels: string[] | null
  created_at: string
  updated_at: string
}

export type CreateKbPermissionRulePayload = {
  name: string
  enabled?: boolean
  user_conditions: UserCondition[]
  scope_operator: ScopeOperator
  scope_labels?: string[] | null
}

export type UpdateKbPermissionRulePayload = Partial<CreateKbPermissionRulePayload>
